### PR TITLE
Tests - added tests for Terms agg partitioning

### PR DESCRIPTION
**Pull request by markharwood**
Added unit tests for testing fromXContent/toXContent implementations of IncludeExclude class. Tests cover "partition" mode plus the regex and exact-value modes of use.
Added a REST test for partitioned terms agg.

Closes https://github.com/elastic/elasticsearch/issues/22102

### CODE REVIEW

**Review by jpountz — APPROVED** [review]
LGTM

Thanks Mark, the new tests look good to me!

Another area that would benefit from testing I think is `FilteredOrdinals` on the one hand, and `acceptedGlobalOrdinals` on the other hand? It is quite unrelated to the test improvements that you did here, so feel free to do it in another PR if that makes things easier to you.

**Comment by markharwood** [comment]
I'm not sure about the REST test though. Failing randomly ("expected 1 got 0", then "expected 1 got 2"...). I assumed hash/modulo of term values would be deterministic regardless of test randomisation but it appears not. Need to dig deeper

**Comment by jpountz** [comment]
Good catch, the seed used to compute the hash of the terms is not stable (`StringHelper.GOOD_FAST_HASH_SEED`). I think we should make term-based partitioning use a static hash and call `StringHelper.murmurhash3_x86_32` manually rather than `terms.hashCode()`. This way we would make sure results are consistent across nodes, and across restarts of a given node?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java [hunk 1]
```diff
@@ -66,6 +66,7 @@
     public static final ParseField PATTERN_FIELD = new ParseField("pattern");
     public static final ParseField PARTITION_FIELD = new ParseField("partition");
     public static final ParseField NUM_PARTITIONS_FIELD = new ParseField("num_partitions");
+    private static final int hash_partitioning_seed = 31;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java]: can you use upper case for consistency and leave a comment about why we set the seed explicitly?

**Review by jpountz — APPROVED** [review]
LGTM

**Comment by markharwood** [comment]
Good to go with this, @s1monw ?

**Review by s1monw — APPROVED** [review]
👍 Awesome

**Comment by markharwood** [comment]
Pushed to master with https://github.com/elastic/elasticsearch/commit/4c6d17a176014fad1bdbe5ed04968455f9435094 and 5.x with https://github.com/elastic/elasticsearch/commit/a06d2cc65f68d63d243cc1c8c883bb9a0325b23f

**Pull request closed without merging**